lib/improver/spotdata/extract_data.py

```diff
@@ -0,0 +1,374 @@
+
+    def process(self, cube, sites, neighbours, ancillary_data, additional_data,
+                no_neighbours=9):
+        """
+        Call the correct function to enact the method of data extraction
+        specified. This function also handles multiple timesteps, consolidating
+        the resulting data cubes into an Iris.CubeList.
+
+        Args:
+        -----
+        cube : iris.cube.Cube
+            Cube of diagnostic data from which to extract spotdata.
+
+        sites : dict
+            A dictionary containing the properties of spotdata sites.
+
+        neighbours : numpy.array
+            Array of neigbouring grid points that are associated with sites
+            in the SortedDictionary of sites.
+
+        additional_data : dict
+            A dictionary containing any supplmentary time varying diagnostics
+            that are needed for the selected extraction method.
+
+        ancillary_data : dict
+            A dictionary containing additional model data that is needed.
+            e.g. {'orography': <cube of orography>}
+
+        no_neighbours : int
+            Number of grid points about the site to consider when calculating a
+            local temperature lapse rate using orography variation.
+            e.g. consider a 5x5 grid of points -> no_neighbours = 25.
+
+        Returns:
+        --------
+        cube : iris.cube.Cube
+            An irregularly (i.e. non-gridded) cube of diagnostic data extracted
+            at the spotdata sites.
+
+        """
+        if self.method == 'use_nearest':
+            return self.use_nearest(cube, sites, neighbours)
+        elif self.method == 'orography_derived_temperature_lapse_rate':
+            orography = data_from_dictionary(ancillary_data, 'orography').data
+            return self.orography_derived_temperature_lapse_rate(
+                cube, sites, neighbours, orography,
+                no_neighbours=no_neighbours)
+        elif self.method == 'model_level_temperature_lapse_rate':
+            height_50 = Constraint(height=50)
+
+            pressure_on_height_levels = (
+                data_from_dictionary(
+                    additional_data, 'pressure_on_height_levels').extract(
+                        height_50)
+                )
+            temperature_on_height_levels = (
+                data_from_dictionary(
+                    additional_data, 'temperature_on_height_levels').extract(
+                        height_50)
+                )
+            surface_pressure = data_from_dictionary(additional_data,
+                                                    'surface_pressure')
+
+            # Ensure that pressure units are in Pa.
+            pressure_on_height_levels.convert_units('Pa')
+            surface_pressure.convert_units('Pa')
+
+            return self.model_level_temperature_lapse_rate(
+                cube, sites, neighbours, pressure_on_height_levels,
+                surface_pressure, temperature_on_height_levels)
+
```

Comment:
There is undoubtedly a larger refactor of all this code to come. I think it has grown to the point where separate classes might be a better approach, rather than a single wrapping class with lots of complex methods within it. I am a little caught by not wanting to get lost in a massive refactor before I get some stuff in the repository. I would like to get spotdata in and then revisit it to improve the structure in conjunction with the wrapping framework a little down the line.